uclibc's gethostbyname can't parse dns udp responce over 512 byte long

diff -BurpN accel-pptp.orig/pppd_plugin/src/pptp.c accel-pptp/pppd_plugin/src/pptp.c
--- accel-pptp.orig/pppd_plugin/src/pptp.c	2009-10-03 09:09:55.000000000 +0000
+++ accel-pptp/pppd_plugin/src/pptp.c	2009-11-06 17:29:42.000000000 +0000
@@ -131,15 +132,37 @@ static int pptp_start_client(void)
 {
 	socklen_t len;
 	struct sockaddr_pppox src_addr,dst_addr;
-	struct hostent *hostinfo;
+	struct hostent *hostinfo = NULL;
+	struct hostent hostentry;
+	size_t buflen = 1024;
+	char *prevbuf, *buffer = NULL;
+	int res, herr;
+
+	while (buflen <= 1024<<4) /* 16Kb should be more than enough */
+	{
+		prevbuf = buffer;
+		if ((buffer = realloc(buffer, buflen)) == NULL)
+		{
+			free(prevbuf);
+			error("PPTP: Can't allocate hostname buffer");
+			return -1;
+		}
+		if ((res = gethostbyname_r(pptp_server, &hostentry, buffer, buflen, &hostinfo, &herr)) == ERANGE)
+		{
+			buflen *= 2;
+			continue;
+		}
+		break;
+	}
 
-	hostinfo=gethostbyname(pptp_server);
-  if (!hostinfo)
+	if (res || !hostinfo)
 	{
+		free(buffer);
 		error("PPTP: Unknown host %s\n", pptp_server);
 		return -1;
 	}
 	dst_addr.sa_addr.pptp.sin_addr=*(struct in_addr*)hostinfo->h_addr;
+	free(buffer);
 
  	memset(&rt, 0, sizeof(rt));
  	route_add(dst_addr.sa_addr.pptp.sin_addr, &rt);
